File: use-cases/school-timetabling/src/main/java/org/acme/schooltimetabling/solver/justifications/RoomConflictJustification.java
package org.acme.schooltimetabling.solver.justifications;

import ai.timefold.solver.core.api.score.stream.ConstraintJustification;
import org.acme.schooltimetabling.domain.Lesson;
import org.acme.schooltimetabling.domain.Room;

public record RoomConflictJustification(Room room, Lesson lesson1, Lesson lesson2) implements ConstraintJustification {

}

File: quickstarts/school-timetabling-kotlin/src/main/kotlin/org/acme/kotlin/schooltimetabling/solver/justifications/RoomConflictJustification.kt
package org.acme.kotlin.schooltimetabling.solver.justifications

import ai.timefold.solver.core.api.score.stream.ConstraintJustification
import org.acme.kotlin.schooltimetabling.domain.Lesson
import org.acme.kotlin.schooltimetabling.domain.Room


data class RoomConflictJustification(
    val room: Room,
    val lesson1: Lesson,
    val lesson2: Lesson,
    val description: String
) :
    ConstraintJustification {

    constructor(
        room: Room,
        lesson1: Lesson,
        lesson2: Lesson
    ) : this(
        room, lesson1, lesson2,
        "Room '%s' is used for lesson '%s' for student group '%s' and lesson '%s' for student group '%s' at '%s %s'"
            .format(
                room,
                lesson1.subject,
                lesson1.studentGroup,
                lesson2.subject,
                lesson2.studentGroup,
                lesson1.timeslot?.dayOfWeek,
                lesson1.timeslot?.startTime
            )
    )
}
